### ISSUE TITLE
[8] Identify bottlenecks in the stack analysis

### ISSUE THREAD

**Issue by tisnik** [openshiftio/openshift.io#1242]
## Description

Update current logging/tooling to be able to identify bottlenecks in the stack analysis

## Acceptance criteria

- [ ] new stack analysis is started, optionally with debug/verbose/ etc. parameter
- [ ] durations for (all) processes and steps can be retrieved via Kibana or via the stack report itself

## Related issues
Task for the epic #1144

**Comment by msrb** [comment]
My findings and observations:

**stack-analysis level**

- it's now possible to retrieve information about tasks that were executed for a particular stack analysis request:
    - for example: `curl -H "Authorization: Bearer $OSIO_TOKEN" https://recommender.api.openshift.io/api/v1/stack-analyses/5e904143e9074248a3294ad58f52da9e/_debug`
- stack analysis consists of multiple tasks, `stack_aggregator_v2` usually takes only couple of millisecond to complete and therefore cost of scheduling it is probably higher than actual execution
- `GraphAggregatorTask` can be slow because we are trying to resolve `pom.xml` files. It's usually blazingly fast for Python stacks as we don't do any resolving there
    - this should be fixed once https://github.com/openshiftio/openshift.io/issues/1063 is in place
- `recommendation_v2` task can take a very long time to finish. But it actually doesn't do any heavy lifting. It synchronously waits for response from PGM
    - response time per request is ~30 seconds on average (see [logs](http://red.ht/2Aps5bl) in Kibana); worker is blocked during this period

I should say that, so far, I've only checked one side of things. We only know what worker thinks - how long it took to process a request in PGM. In order to rule out possible network issues in the cluster and misconfiguration of the PGM's web server (maybe requests actually spend most of the time in a backlog), we need to also find out what PGM thinks, i.e. measure the times directly in the PGM service as well. This is yet to be done.

I think it would be worth it to rethink and redesign stack analysis flow. Currently the biggest bottleneck seems to be PGM (but see the previous paragraph) and the fact that we are synchronously waiting for the response, effectively blocking workers which could be processing other requests in the meantime.

There was also missing index in RDS, on a column that we use for retrieving worker results. This is already fixed and it could be the reason behind some 50x status codes returned randomly by the API.

Low hanging fruit:
- for stack-analyses GET request, we are doing 3 separate calls to RDS to retrieve worker results; one should do the trick


TL;DR:
- stack analyses cannot really scale if all available workers are blocked for long period of time. Response times on PGM side should be measured first and the flow should be redesigned, if needed.

**Comment by msrb** [comment]
Example response from the new debug endpoint:
```bash
$ curl -H "Authorization: Bearer $OSIO_TOKEN" https://recommender.api.openshift.io/api/v1/stack-analyses/5e904143e9074248a3294ad58f52da9e/_debug
{
  "tasks": [
    {
      "ended_at": "2017-11-06T08:01:51.882315", 
      "error": false, 
      "started_at": "2017-11-06T08:01:11.193871", 
      "task_name": "recommendation_v2"
    }, 
    {
      "ended_at": "2017-11-06T08:00:42.476830", 
      "error": false, 
      "started_at": "2017-11-06T08:00:42.348812", 
      "task_name": "stack_aggregator_v2"
    }, 
    {
      "ended_at": "2017-11-06T08:00:35.623950", 
      "error": false, 
      "started_at": "2017-11-06T08:00:31.657956", 
      "task_name": "GraphAggregatorTask"
    }
  ]
}

```

**Comment by miteshvp** [comment]
@msrb - good findings. It will definitely help to identify PGM scoring rate. We did perform PGM load tests previously, but lot has changed over time and it asks for one more round of load tests. I've created an issue which can be tracked for PGM load tests - https://github.com/openshiftio/openshift.io/issues/1306

**Comment by msrb** [comment]
Just FYI, response time logging in kronos: https://github.com/fabric8-analytics/fabric8-analytics-stack-analysis/pull/123

**Comment by msrb** [comment]
Follow up task: https://github.com/openshiftio/openshift.io/issues/1438
